### ISSUE TITLE
MATERIAL DESIGN: As an OC user I want that the new list & grid components from Material Design replace the current so that performance is improved in folders with hundreds of files & images

### ISSUE THREAD

**Issue by davivel** [owncloud/android#984]
Google states that the new elements for lists and grid view included in the Android Support Library help to improve the performance of lists and grids with tons of elements. I couldn't check it, but it's obvious they *enforce* the recycling of views instead of simply allowing the option to do it (or the option to forget to do it and have problems; see #955 fixing it). If there is also a better recycling policy behind, we will only notice it by start using them.

Nevertheless, there are other important reasons to embrace this change:

- The GridView widget currently in use is a mess; it's not fully implemented in all the Android versions (even in different 4,x versions). This fact already caused crashes in the past, and is still enforcing some undesired graphical effect in older 4.x devices (my lips are sealed). Using the new Grid element in Android Support Library we will have a single and complete implementation for all our grids.

- Both components are needed if we want to take advantage of new elements for cool animations in 

Sure of interest for @tobiasKaminsky , @jancborchardt , and everybody is welcome.

Just please, remember, that #692  should be done first (or at least the part of the build target).

**Comment by masensio** [comment]
Grid lists and lists from Material Design
http://www.google.com/design/spec/components/grid-lists.html#
http://www.google.com/design/spec/components/lists.html

**Comment by skymania** [comment]
why just not use recycler view and LayoutManager to show list or Grid?

**Comment by Kernald** [comment]
@skymania actualy, that's the whole point of this issue.

**Comment by skymania** [comment]
I hope so.

**Comment by jancborchardt** [comment]
@skymania since it sounds like you’re experienced, maybe you’re interested in contributing? :) I’m sure if you need any help, @davivel @tobiasKaminsky @AndyScherzinger can help you.

**Comment by AndyScherzinger** [comment]
Absolutely! :) The material_toolbar branch is probably a good starting point since it contains most of the material implementation and the latest AppCompat release.

**Comment by skymania** [comment]
I have implemented toolbar, recycler view and UIL. Currently Im working on navigation drawer but time is not currently on my side so I have no idea when I will have the time to proceed with material implenetations. I have to upload all the changes I made on github tho.

**Comment by jancborchardt** [comment]
@skymania please do a pull request even when it's work in progress so we can already review. If a pull request is too big, risk is high it's very difficult to merge.

**Comment by AndyScherzinger** [comment]
@ all how should we all move forward with this overall topic, since we now have this, #1076 and #1090 and @skymania provides the next implementation (which is fine with me) and I provided the action bar variant. So imho we need some kind of coordination to not implement everything twice :)

**Comment by jancborchardt** [comment]
I’d say we should wait for #1076 to be QA'd and merged – which will hopefully happen soon. Then #1090 shortly after (we need it in the next cycle @davivel).

Then after that we can build on top of that. Otherwise it just becomes unmanageable. There’s no reason to cram all of these improvements in such a short amount of time now. ;)

**Comment by AndyScherzinger** [comment]
Sounds reasonable @jancborchardt - thanks for the clarification :)

**Comment by skymania** [comment]
Finally found some time and I am currently using this for my owncloud server. Still in development tho. 

![0](https://cloud.synox.si/index.php/s/joYbOl1rRwfTbZH/download)

![1](https://cloud.synox.si/index.php/s/SNayLPGggLyqD6R/download)

![2](https://cloud.synox.si/index.php/s/B5msBXEVbP2UUZL/download)

![3](https://cloud.synox.si/index.php/s/sGBXxNOavrtgufC/download)

![4](https://cloud.synox.si/index.php/s/iH99R0skPNI56bA/download)

**Comment by AndyScherzinger** [comment]
@skymania did you rewrite one of the material branches of the ownCloud Android app?! :question:

**Comment by skymania** [comment]
the master one.

**Comment by AndyScherzinger** [comment]
hmm, okay. The fab implementation would have been on the material_fab branch already ;)

**Comment by AndyScherzinger** [comment]
Any news on this matter? @skymania fancy some collaboration?

I would start working on this and will create a new branch for the recycler view implementation based on the actual master a.k.a soon to be v2.0.0 all formerly mentioned PRs and issues have been merged or are at least done from a implementation point of view only missing out on review/test/merge.

I will update the issue referencing all the issues.

Branch has been created: https://github.com/owncloud/android/tree/984_recycler_view 
So who every wants to work with me on this, feel free to do so either via direct commits to the branch or PRs targeting ```984_recycler_view ```

**Comment by AndyScherzinger** [comment]
also looping in @malkomich since this is an issue raised in 2015 :)

**Comment by skymania** [comment]
@AndyScherzinger what ever you need you can take from Cirrus.
It already has a recycler view implementation.

**Comment by AndyScherzinger** [comment]
That would be awesome! :+1: 

To thing though, just to be save:
1. Could you sign the contributor agreement https://owncloud.org/contribute/agreement/ ? (If I re-use your implementation and not do it from scratch on my own the code couldn't be merged)
2. Would you do code reviews on my changes / your"copied" code?

**Comment by davivel** [comment]
@skymania , any chance you make a PR from your fork to ownCloud with the changes for this topic?

**Comment by davivel** [comment]
@AndyScherzinger , I disabled checkbox on #1159 since it's not merged yet; was a bit confusing. Hope it's OK to you.

**Comment by AndyScherzinger** [comment]
Sure, you mean 1559 (?) I guess (drawer) which has just been mentioned because the discussion in this issue simply mentioned drawer, so I referenced it to explicitly document that that part has already been done.

**Comment by davivel** [comment]
That was, I fixed the comment.

**Comment by skymania** [comment]
@AndyScherzinger  so I sign contribution agreement and we can start? I would love to contribute and do code reviews, but just to warn you, we will have to do shit loads of changes to get this right :D

**Comment by davivel** [comment]
That's nice, @skymania .

If you had to do a summary, what would those lot of changes include?

**Comment by skymania** [comment]
@davivel 
- implementation of recycler view
- search 
- totally changed image loading mechanism (the current one is blocking UI)
- avoiding multiple calls to notifiyDataSetChanged (this is happening now for loading content at start and then for every call we get back from broadcast recevier and its blocking UI in folders with lot of contents, for example folders with lots of images)
- async fill and notify adapter that the data has changed 

and probably something else that I have forgot now  :)

**Comment by AndyScherzinger** [comment]
@skymania yes that is basically it :) After signing the agreement and sending it to @karlitschek you will get the corresponding access rights to the repo and we can start working on a branch in the owncloud repo itself :)

I am fully aware that it is a huge amount of work that needs to be done, probably larger than what you had to do for cirrus since we do and (and will have) to more requirements for the lists:
* multi select support for @tobiasKaminsky's change
* header and footer e.g. for the new uploads view introduced in v.2.0.0

I am totally in line with your list that will need to be implemented. I also think we will need an abstract implementation for reuse throughout the lists that is able to do the things you and I mentioned.

Really looking forward to the collaboration and finally bringing RecyclerView to the app :+1:

**Comment by AndyScherzinger** [comment]
The agreement can be found here: https://owncloud.org/contribute/agreement/
When I signed it I just sent a scan, which was totally fine :)

**Comment by skymania** [comment]
@AndyScherzinger done. 
Lets do some damage :)

**Comment by davivel** [comment]
Sounds great, guys.

But please, don't forget that we need to keep an order so that process is manageable. All those changes shouldn't go in a single branch. Each of them should go in a different branch created from master, and target master when done in separate PRs.

I'm eager to see how this progresses :)

**Comment by skymania** [comment]
@davivel I'll leave making seperate branches to @AndyScherzinger, he knows how the work is done here more than I do. :)

**Comment by AndyScherzinger** [comment]
Sure thing, that isn't to much of an issue I can create as many branches as necessary :+1: 
About the splitting though, I am not sure to which extend we can split it, since a basic recycler implementation would already mean a certain amount of changes. So we could then for example open separate branches for:
* rudimentary recycler view implementation
* header/footer
* multi select
* optimized image loading
* search

**Comment by skymania** [comment]
Are there any reason the 2.0.0 version is still using ActionBar instead of Toolbar implemetation?

**Comment by AndyScherzinger** [comment]
Yes, that is due to the fact that #1559 hasn't yet been merged back to master and thus wasn't released yet.

**Comment by skymania** [comment]
@AndyScherzinger there I made PR on the branch. Its a start so many functions will not work properly, but basic navigation is implemented. We can go on on from there.

**Comment by karlitschek** [comment]
@AndyScherzinger  @skymania  Confirmed :-)

**Comment by AndyScherzinger** [comment]
@skymania sound good to me! Great to have you on board! :+1:

**Comment by skymania** [comment]
Thnx mate :)

**Comment by AndyScherzinger** [comment]
@skymania you should now have write access to the owncloud repo. I would therefor suggest that we just merge your actual PR and that you then switch to a clone of the owncloud repo branch and we simply work on that one and directly push to that one :)

**Comment by skymania** [comment]
@AndyScherzinger sounds good :+1:

**Comment by AndyScherzinger** [comment]
@skymania I just rebased to the latest master to resolve some conflicts with the latest build changes (removal of maven support). So you need to update before doing anymore commits :)

**Comment by davivel** [comment]
Sorry, but we need to drop issues from milestone 2.4.0 to get it out next to OC 10 server.

This time we'll move topics to ```backlog``` instead of directly to next release, so that we can schedule 2.5.0 properly.

**Comment by DeepDiver1975** [comment]
🤖 Automated issue triage by Claude Code review agent.

This issue was opened before 2020 and has had no activity since then, so we are
closing it as outdated. If this is still relevant on a current ownCloud release,
please **reopen** (or open a fresh issue) with up-to-date details — thank you\!
